### PR TITLE
feat: support S3 Express One Zone

### PR DESCRIPTION
**Pull request by lauzadis**
Adds some features in order to support the SDK making S3 Express requests

## Description of changes

- Refactor checksums to use an execution context key `HttpOperationContext.ChecksumAlgorithm`. The value is the name of the checksum algorithm to be used. If it's set when flexible checksums runs, it will be used, otherwise the `Content-MD5` header will be computed and added as before. This greatly simplifies the logic required in these interceptors.
- Allow setting `omitSessionToken` signing configuration via execution context, using the `AwsSigningAttributes. OmitSessionToken` key.
- Add a generic, thread-safe `LruCache` which holds entries up to a configurable `capacity` and then evicts the least-recently used entries as more are added.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    public suspend fun <T> withLock(block: suspend () -> T): T = mu.withLock {
+        block()
+    }
+
+    /**
+     * Get the set of entries in the cache.
+     * Note: This is not thread safe! If you plan to iterate over entries, lock the mutex.
+     */
+    public val entries: MutableSet<MutableMap.MutableEntry<K, V>>
+        get() = map.entries
```
Comment (ianbotsf) [inline runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt:87]: **Question**: Providing mutable, live access to a volatile underlying collection seems dangerous. Why is this necessary? Would it serve the same purpose to provide an immutable snapshot of the entries? Or even of the entire map?

```kotlin
public class LruCache<K, V>(public val capacity: Int) {
    private val mu = Mutex()
    private val map = linkedMapOf<K, V>()
    …
    public val snapshot: Map<K, V>
        get() = map.toMap()
    …
}
```

**Review by ianbotsf — CHANGES_REQUESTED** [review]

**Review by aajtodd — CHANGES_REQUESTED** [review]

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt [hunk 2]
```diff
@@ -294,6 +294,9 @@ internal class AuthHandler<Input, Output>(
             // update the request context with endpoint specific auth signing context
             val endpointAuthAttributes = endpoint.authOptions.firstOrNull { it.schemeId == authScheme.schemeId }?.attributes ?: emptyAttributes()
             request.context.merge(endpointAuthAttributes)
+
+            // also update the request context with endpoint attributes
+            request.context.merge(endpoint.attributes)
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt:299]: Without this change, `AttributeKey(backend)=S3Express` is not set in the execution context. Should it be coming in from the `endpointAuthAttributes`? If so, there may be a bug here somewhere

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * A thread-safe generic LRU (least recently used) cache.
+ * Entries will be added up to a configured [capacity].
+ * Once full, adding a new entry will evict the least recently used entry.
+ */
+public class LruCache<K, V>(
+    public val capacity: Int,
+) {
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt:17]: It could be, I'll update it.

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * Lock the mutex and execute a function [block].
+     * Note: It is safe for [block] to call [getUnlocked] and [putUnlocked].
+     * @param block A function accepting zero arguments and returning some value [T]
+     * @return The result of the function, [T]
+     */
+    public suspend fun <T> withLock(block: suspend () -> T): T = mu.withLock {
+        block()
+    }
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt:80]: This method was required in order to safely do async cache refresh in S3ExpressCredentialsCache. Without this, the cache may be updated in the middle of being refreshed which would cause issues

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt (hunk 1 above)
Comment (lauzadis) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt:87]: Yes, I was not intending for this to be mutable, I'll update it

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt [hunk 5]
```diff
@@ -29,23 +30,13 @@ import kotlin.coroutines.coroutineContext
...
 @InternalApi
-public class FlexibleChecksumsRequestInterceptor<I>(
-    private val checksumAlgorithmNameInitializer: (I) -> String?,
-) : HttpInterceptor {
-    private var checksumAlgorithmName: String? = null
-
-    override fun readAfterSerialization(context: ProtocolRequestInterceptorContext<Any, HttpRequest>) {
-        @Suppress("UNCHECKED_CAST")
-        val input = context.request as I
-        checksumAlgorithmName = checksumAlgorithmNameInitializer(input)
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt:44]: This change now allows interceptors to override the checksum algorithm for a request. It's necessary to enable CRC32 for S3 Express and may be useful later, do you think it's problematic?

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt [hunk 6]
```diff
@@ -29,23 +30,13 @@ import kotlin.coroutines.coroutineContext
...
-    }
+public class FlexibleChecksumsRequestInterceptor : HttpInterceptor {
+    override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        val logger = coroutineContext.logger<FlexibleChecksumsRequestInterceptor>()
 
-    override suspend fun modifyBeforeRetryLoop(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt:47]: This may be leftover from my original interpretation of the SEP's checksum rules (which was wrong), I'll take a look and see if I can revert it.

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt [hunk 7]
```diff
@@ -29,23 +30,13 @@ import kotlin.coroutines.coroutineContext
...
-        val logger = coroutineContext.logger<FlexibleChecksumsRequestInterceptor<I>>()
+        val checksumAlgorithmName = context.executionContext.getOrNull(HttpOperationContext.ChecksumAlgorithm)
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt:39]: Yeah, I wanted to have a discussion about these interceptor changes + how they affect binary compatibility. I'll take a look at the second option since it seems better

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt:299]: https://github.com/awslabs/smithy-kotlin/pull/1033#discussion_r1486834290

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (aajtodd) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt:80]: Can you elaborate, I agree with Ian here on this that generally we want to avoid exposing implementation details. If something is documented as thread safe it shouldn't need to expose it's synchronization internals. It feels like we're forcing external requirements unrelated to the cache into the implementation/interface of this type.

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (aajtodd) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt:299]: I think I see. It would appear we merge the signing attributes which is the extracted endpoint auth attributes for the chose auth scheme. This may be the first case where there are other endpoint attributes unrelated to auth?

It sure feels like this belongs to auth attributes still since it's related to the S3 express auth scheme right? What does the endpoint rule look like here?

[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt:299]: It looks like this:
```
    "endpoint": {
        "url": "{url#scheme}://{url#authority}/{uri_encoded_bucket}{url#path}",
        "properties": {
            "backend": "S3Express",
            "authSchemes": [{
                "disableDoubleEncoding": true,
                "name": "sigv4",
                "signingName": "s3express",
                "signingRegion": "{Region}"
            }]
        },
        "headers": {}
    }
```

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt:47]: Reminded myself... it's because the endpoint is resolved after `modifyBeforeRetryLoop`. S3 Express needs the checksum set to CRC32 only if `backend=s3express` (endpoint attribute) and that isn't known at the `modifyBeforeRetryLoop` stage. 

So, I had to push back all these checksum-related interceptors to give S3 Express a chance to set CRC32 before flexible checksums / MD5 checksum runs.

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt [hunk 8]
```diff
@@ -21,22 +22,9 @@ import aws.smithy.kotlin.runtime.text.encoding.encodeBase64String
...
-        } ?: true
-    }
-
-    override suspend fun modifyBeforeRetryLoop(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
-        if (!shouldInjectMD5Header) {
+public class Md5ChecksumInterceptor : HttpInterceptor {
+    override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        if (context.executionContext.contains(HttpOperationContext.ChecksumAlgorithm)) {
             return context.protocolRequest
         }
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt]: I think I found a good solution here. The Md5ChecksumInterceptor no longer explicitly checks the execution context. Instead, the execution context is checked in the (poorly-named, now documented) block. This means non-AWS users can pass a custom `block` that doesn't consider the execution context `ChecksumAlgorithm` and they can have both MD5 _and_ flexible checksums

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt]: I ended up going with the first option because the `FlexibleChecksumsResponseInterceptor` still takes the input type param `<I>`

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt:47]: We spoke offline and agreed this is fine as long as the checksum is cached so it's not recomputed every attempt. 

That's not easy to do for the `deferredChecksum` (chunked) route because that checksum is only done being calculated once the interceptor is done executing. I added caching for the non-chunked request checksums.

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt [hunk 9]
```diff
@@ -21,22 +22,9 @@ import aws.smithy.kotlin.runtime.text.encoding.encodeBase64String
...
-            @Suppress("UNCHECKED_CAST")
-            val input = context.request as I
-            it(input)
-        } ?: true
-    }
-
-    override suspend fun modifyBeforeRetryLoop(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
-        if (!shouldInjectMD5Header) {
+public class Md5ChecksumInterceptor : HttpInterceptor {
+    override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt:31]: Spoke offline and agreed this is fine as long as we cache the checksum so it's not recomputed every time. That's what happens now.

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * Add or update a cache entry with a key [k] and value [v].
+     * @param k the key to associate the value with
+     * @param v the value to store in the cache
+     * @return [Unit]
+     */
+    public suspend fun put(k: K, v: V): Unit = mu.withLock {
+        if (k !in map && map.size == capacity) {
+            map.remove(map.keys.first())
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt:41]: On JVM, [LinkedHashMap](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-linked-hash-map/) is typealiased to [Java's LinkedHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html), which provides constant-time access to the first and last keys since it's backed by a doubly linked list

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+     * Remove an entry associated with a key [k], if it exists.
+     * @param k the key to remove from the cache
+     * @return the value removed, or null if it did not exist
+     */
+    public suspend fun remove(k: K): V? = mu.withLock { map.remove(k) }
+
+    /**
+     * Get a snapshot of the entries in the cache.
+     */
+    public val entries: Set<Map.Entry<K, V>>
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt:57]: They are used in the async refresh of S3ExpressCredentialsCache. I added a comment that they're not thread-safe. But these are general methods that most collections offer, so I think we should keep them.

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt [hunk 12]
```diff
@@ -19,30 +19,30 @@ import aws.smithy.kotlin.runtime.text.encoding.encodeBase64String
...
  *   - https://awslabs.github.io/smithy/1.0/spec/core/behavior-traits.html#httpchecksumrequired-trait
  *   - https://datatracker.ietf.org/doc/html/rfc1864.html
+ * @param block An optional function which parses the input [I] to determine if the `Content-MD5` header should be set.
+ * If not provided, the default behavior will set the header.
  */
 @InternalApi
 public class Md5ChecksumInterceptor<I>(
     private val block: ((input: I) -> Boolean)? = null,
 ) : HttpInterceptor {
+    private var cachedChecksum: String? = null
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt:29]: I like the idea, but I think it would be an API break. Users making custom `Md5ChecksumInterceptor` / `FlexibleChecksumsRequestIntereceptor` would need to implement these new methods (`calculateChecksum` / `applyChecksum`).

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt [hunk 13]
```diff
@@ -92,17 +97,18 @@ public class FlexibleChecksumsRequestInterceptor<I>(
...
-            val checksum: String = when {
-                req.body.contentLength == null && !req.body.isOneShot -> {
-                    val channel = req.body.toSdkByteReadChannel()!!
-                    channel.rollingHash(checksumAlgorithm).encodeBase64String()
-                }
-                else -> {
-                    val bodyBytes = req.body.readAll()!!
-                    req.body = bodyBytes.toHttpBody() // replace the consumed body
-                    bodyBytes.hash(checksumAlgorithm).encodeBase64String()
+            val checksum: String = cachedChecksum ?: run {
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt:100]: refactored into `calculateChecksum`

**Review by ianbotsf — APPROVED** [review]

File: runtime/auth/aws-signing-crt/jvm/src/aws/smithy/kotlin/runtime/auth/awssigning/crt/CrtAwsSigner.kt [hunk 14]
```diff
@@ -26,15 +26,24 @@ public object CrtAwsSigner : AwsSigner {
...
+        val isS3Express = request.headers.contains("X-Amz-S3session-Token")
+
+        val requestBuilder = request.toBuilder()
 
-        val crtResult = CrtSigner.sign(crtRequest, crtConfig)
+        val crtConfig = config.toCrtSigningConfig().toBuilder()
+        if (isS3Express) {
+            crtConfig.algorithm = CrtSigningAlgorithm.SIGV4_S3EXPRESS
+            crtConfig.omitSessionToken = false
+            requestBuilder.headers.remove("X-Amz-S3session-Token")
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/auth/aws-signing-crt/jvm/src/aws/smithy/kotlin/runtime/auth/awssigning/crt/CrtAwsSigner.kt:37]: Yes it does re-add this header, and it fails if the header is already present

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt:29]: Good point, I didn't fully think this through... I can try the refactor and see how it looks

File: runtime/auth/aws-signing-crt/jvm/src/aws/smithy/kotlin/runtime/auth/awssigning/crt/CrtAwsSigner.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (ianbotsf) [reply runtime/auth/aws-signing-crt/jvm/src/aws/smithy/kotlin/runtime/auth/awssigning/crt/CrtAwsSigner.kt:37]: Got it. Let's add a comment to that effect.

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt]: FYI: I did the refactor and the .api file appears "broken" because the implementation changed from `HttpInterceptor` to `AbstractChecksumInterceptor` but under the hood it's still valid

File: runtime/auth/aws-signing-crt/jvm/src/aws/smithy/kotlin/runtime/auth/awssigning/crt/CrtAwsSigner.kt [hunk 15]
```diff
@@ -26,15 +26,24 @@ public object CrtAwsSigner : AwsSigner {
     override suspend fun sign(request: HttpRequest, config: AwsSigningConfig): AwsSigningResult<HttpRequest> {
         val isUnsigned = config.hashSpecification is HashSpecification.UnsignedPayload
         val isAwsChunked = request.headers.contains("Content-Encoding", "aws-chunked")
-        val crtRequest = request.toSignableCrtRequest(isUnsigned, isAwsChunked)
-        val crtConfig = config.toCrtSigningConfig()
+        val isS3Express = request.headers.contains("X-Amz-S3session-Token")
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply runtime/auth/aws-signing-crt/jvm/src/aws/smithy/kotlin/runtime/auth/awssigning/crt/CrtAwsSigner.kt:29]: It's only used once in **smithy-kotlin**. It's already a constant in **aws-sdk-kotlin** but it's not referencable here

**Review by aajtodd — APPROVED** [review]
